File: src/node/c_api/folder.cpp
/***

  Oak Video Editor - Non-Linear Video Editor
  Copyright (C) 2026 Oak Team

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "node/folder.h"

#include "../../undo/c_api/commandhandle.h"

#include <new>

#include "node/node.h"

#include "../src/project.h"
#include "../src/project/folder/folder.h"

#include "nodehandle.h"

using oaknode_c_api::delete_as;
using oaknode_c_api::make_handle;
using oaknode_c_api::to_native;

OakNodeFolder oaknode_folder_create(OakNodeProject project)
{
	if (!project.ctx) {
		return OakNodeFolder{};
	}

	try {
		auto *folder = new (std::nothrow) olive::Folder();
		if (!folder) {
			return OakNodeFolder{};
		}
		to_native<olive::Project>(project)->add_node(folder);
		// Borrowed handle: the project graph owns the folder.
		return make_handle<OakNodeFolder>(folder, false,
										  &delete_as<olive::Folder>);
	} catch (...) {
		return OakNodeFolder{};
	}
}

int oaknode_folder_child_count(OakNodeFolder folder)
{
	if (!folder.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		return to_native<olive::Folder>(folder)->item_child_count();
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

OakNodeNode oaknode_folder_child_at(OakNodeFolder folder, int index)
{
	if (!folder.ctx || index < 0 ||
		index >= to_native<olive::Folder>(folder)->item_child_count()) {
		return OakNodeNode{};
	}

	try {
		return make_handle<OakNodeNode>(
			to_native<olive::Folder>(folder)->item_child(index), false,
			&delete_as<olive::Node>);
	} catch (...) {
		return OakNodeNode{};
	}
}

int oaknode_folder_add_child(OakNodeFolder folder, OakNodeNode child)
{
	if (!folder.ctx || !child.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		olive::Folder *f = to_native<olive::Folder>(folder);
		olive::Node *c = to_native<olive::Node>(child);

		if (c->folder()) {
			return OAKNODE_E_STATE;
		}

		olive::FolderAddChild cmd(f, c);
		cmd.redo_now();
		// The graph now owns the child; releasing `child` must not
		// delete it.
		oaknode_c_api::mark_container_owned(child);
		return OAKNODE_OK;
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

int oaknode_folder_remove_child(OakNodeFolder folder, OakNodeNode child)
{
	if (!folder.ctx || !child.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		olive::Folder *f = to_native<olive::Folder>(folder);
		olive::Node *c = to_native<olive::Node>(child);

		if (f->index_of_child(c) == -1) {
			return OAKNODE_E_NOT_FOUND;
		}

		olive::Folder::RemoveElementCommand cmd(f, c);
		cmd.redo_now();
		return OAKNODE_OK;
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

int oaknode_folder_move_children(const OakNodeNode *nodes, int count,
								 OakNodeFolder dest_folder)
{
	if (!nodes || count < 0 || !dest_folder.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		olive::Folder *dest = to_native<olive::Folder>(dest_folder);

		for (int i = 0; i < count; i++) {
			if (!nodes[i].ctx) {
				return OAKNODE_E_INVALID;
			}

			olive::Node *node = to_native<olive::Node>(nodes[i]);
			olive::Folder *old_folder = node->folder();

			if (old_folder == dest) {
				continue;
			}

			if (old_folder) {
				olive::Folder::RemoveElementCommand remove_cmd(old_folder, node);
				remove_cmd.redo_now();
			}

			olive::FolderAddChild add_cmd(dest, node);
			add_cmd.redo_now();
			// The graph owns the moved node; releasing the caller's
			// handle must not delete it.
			oaknode_c_api::mark_container_owned(nodes[i]);
		}

		return OAKNODE_OK;
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

int oaknode_folder_has_child_recursive(OakNodeFolder folder,
									   OakNodeNode child)
{
	if (!folder.ctx || !child.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		return to_native<olive::Folder>(folder)->has_child_recursive(
				   to_native<olive::Node>(child))
				   ? 1
				   : 0;
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

int oaknode_folder_index_of_child(OakNodeFolder folder,
								  OakNodeNode child)
{
	if (!folder.ctx || !child.ctx) {
		return OAKNODE_E_INVALID;
	}

	try {
		int index = to_native<olive::Folder>(folder)->index_of_child(
			to_native<olive::Node>(child));
		return index == -1 ? OAKNODE_E_NOT_FOUND : index;
	} catch (...) {
		return OAKNODE_E_FAILED;
	}
}

OakNodeFolder oaknode_folder_parent_of(OakNodeNode node)
{
	if (!node.ctx) {
		return OakNodeFolder{};
	}

	try {
		return make_handle<OakNodeFolder>(
			to_native<olive::Node>(node)->folder(), false,
			&delete_as<olive::Folder>);
	} catch (...) {
		return OakNodeFolder{};
	}
}

OakNodeNode oaknode_folder_as_node(OakNodeFolder folder)
{
	// Borrowed cast: same object, the handle only releases itself.
	return make_handle<OakNodeNode>(to_native<olive::Folder>(folder), false,
									&delete_as<olive::Node>);
}

OakUndoCommand oaknode_command_create_folder_add_child(
	OakNodeFolder folder, OakNodeNode child)
{
	if (!folder.ctx || !child.ctx) {
		return OakUndoCommand{};
	}

	try {
		return oakundo_capi::make_command_handle(
			new olive::FolderAddChild(to_native<olive::Folder>(folder),
									  to_native<olive::Node>(child)),
			true);
	} catch (...) {
		return OakUndoCommand{};
	}
}

File: crates/oaknode/src/folder.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Bin folders (C++ `olive::Folder`): tree structure over `NodeId`s.

use crate::factory::NodeMeta;
use crate::id::NodeId;
use crate::node::{Category, NodeBehavior, NodeCore};

/// A folder node behavior: children are folder-tree members (folders
/// and footage), not graph edges.
pub struct FolderBehavior {
	/// Display name.
	pub name: String,
	/// Child node ids (folders/footage) in bin order.
	pub children: Vec<NodeId>,
}

impl FolderBehavior {
	/// New empty folder.
	pub fn new(name: &str) -> Self {
		FolderBehavior {
			name: name.to_string(),
			children: Vec::new(),
		}
	}

	/// Add a child (C++ `Folder::add_child`).
	pub fn add_child(&mut self, child: NodeId) {
		if !self.children.contains(&child) {
			self.children.push(child);
		}
	}

	/// Remove a child; false when absent.
	pub fn remove_child(&mut self, child: NodeId) -> bool {
		let before = self.children.len();
		self.children.retain(|c| *c != child);
		self.children.len() != before
	}

	/// Index of `child` in the direct children.
	pub fn index_of_child(&self, child: NodeId) -> Option<usize> {
		self.children.iter().position(|c| *c == child)
	}

	/// True when `child` is contained recursively (C++
	/// `Folder::has_child_recursive`).
	pub fn has_child_recursive(&self, child: NodeId, graph: &crate::graph::Graph) -> bool {
		for c in &self.children {
			if *c == child {
				return true;
			}
			if let Some(entry) = graph.get(*c) {
				if let Some(f) = entry
					.behavior
					.as_any()
					.and_then(|a| a.downcast_ref::<FolderBehavior>())
				{
					if f.has_child_recursive(child, graph) {
						return true;
					}
				}
			}
		}
		false
	}
}

impl NodeBehavior for FolderBehavior {
	/// Human-readable name (C++ `name()`): the folder's display name.
	fn name(&self) -> &str {
		&self.name
	}

	/// Concrete downcast for the folder family.
	fn as_any(&self) -> Option<&dyn std::any::Any> {
		Some(self)
	}

	/// Concrete downcast for the folder family.
	fn as_any_mut(&mut self) -> Option<&mut dyn std::any::Any> {
		Some(self)
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.folder"
	}

	/// Categories (C++ `category()`): items live in the bin, not the
	/// render graph.
	fn categories(&self) -> &[Category] {
		&[Category::Timeline]
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, _core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		Some(Box::new(FolderBehavior::new(&self.name)))
	}

	/// Custom project save: the bin children (C++ attaches children
	/// through the `child_in` input connections; the Rust model keeps
	/// them in [`FolderBehavior::children`], so the custom segment is
	/// the persistence channel — old readers skip it).
	fn save_custom(&self, core: &NodeCore, writer: &mut dyn crate::serializer::XmlWrite) {
		let _ = core;
		if !self.children.is_empty() {
			writer.start_element("children");
			for c in &self.children {
				writer.text_element("child", &c.identity().to_string());
			}
			writer.end_element(); // children
		}
	}

	/// Custom project load; the child references resolve in the
	/// serializer's post-load pass (which also folds in `child_in`
	/// connections from C++ files).
	fn load_custom(
		&mut self,
		_core: &mut NodeCore,
		reader: &mut dyn crate::serializer::XmlRead,
	) -> bool {
		while reader.next_start_element() {
			match reader.name() {
				"children" => {
					self.children.clear();
					while reader.next_start_element() {
						if reader.name() == "child" {
							if let Some(id) =
								crate::serializer::parse_node_ref(&reader.read_element_text())
							{
								self.children.push(id);
							}
						} else {
							reader.skip_current_element();
						}
					}
				}
				_ => reader.skip_current_element(),
			}
		}
		true
	}
}

/// Constructor (C++ `Folder::Folder()`): a folder node declares the
/// `child_in` array input (C++ attaches bin children through it) but no
/// `enabled_in` (`// CPP-PARITY: folder.cpp:26`).
pub fn create(name: &str) -> (NodeCore, Box<dyn NodeBehavior>) {
	let mut core = NodeCore::empty();
	let mut child = crate::input::Input::new(
		"child_in",
		crate::value::ValueType::None,
		crate::value::NodeValue::None,
	);
	child.flags |= crate::input::flags::ARRAY | crate::input::flags::NOT_KEYFRAMABLE;
	core.add_input(child);
	(core, Box::new(FolderBehavior::new(name)))
}

/// Register a folder-typed node (used by the serializer for bin folders;
/// folders are not in the factory menu).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.folder",
		name: "Folder",
		categories: &[Category::Timeline],
		create: || create("Folder"),
	});
}
